### DISCUSSION TITLE
YOLOv8 Model in SAHI with “no_sliced_prediction=True”  got 0.26 less mAP50 than before？

### DISCUSSION THREAD

**Discussion by WaderLaken** [obss/sahi#1030] [Q&A] [answered]
First, thank you for your contributions to SAHI.
But I get an incomprehensible problem using SAHI with YOLOv8 model. 

When I compute the metrics for val datesets using **standard inference** with a YOLOv8 Model, I could get **0.862 mAP50**. However, I just get **0.607 mAP50** using **sliced inference** with the same model, simultaneously setting "no_sliced_prediction=True" or setting "no_sliced_prediction=False" but tuning the "slice_size" to an oversized value, like 8000x8000, definitively inferring only one slice on a picture. By the way, I ensure the other reasoning variables the same. What a surprising result!

What's more, I make another try, using sliced inference with am appropriate slice-size (such as 4032x4032), and the **mAP50 rises to 0.931**. Now I confirm that SAHI is effective to improve the recognition accuracy, especially for tiny objects. However, I really wonder why the values between the last two experiments are markedly different?

Looking forward to your reply. @sinanonur @fcakyon
Best wishes!

**Comment by sinanonur** [comment]
Hello @WaderLaken 

I suspect mAP 's not calculated with the same library but It needs to be checked.
Using "standard inference" how do you calculate mAP50  and is it the same with SAHI prediction?  (different libraries may yield different results)

Also can you make sure there is no post processing with your code?

You can also compare the results of no sliced prediction with standard inference. Also you can try to calculate the mAP in a  They should be the same. It not it might require further investigation.

If prediction results are different Is it possible for you to share a notebook or code with a sample image demonstrating this situation?


Best

**Reply by WaderLaken** [reply]
Glad to hear from you @sinanonur 

I will show you more effective details.

To be honest, I'm certain that I use the same library `sahi/scripts/coco_evaluation.py` when I calculate mAP for all inferences. And the predictions from standard YOLOv8 are converted from yolo to coco format before calculating. What's more, the inferences either using standard YOLOv8 or SAHI are based on the same setup, such as 0.001 conf, 0.5 iou, 1 batch-size, 640 image-size, 1000 max-detections and so on.

Also there has been no post processing added or modified with my code yet.

Well, to make it more intuitive, I list the results as follows:
1. standard YOLOv8:  **`0.862 mAP50`** (converted from yolo to coco, then fed into `sahi/scripts/coco_evaluation.py`)
2. SAHI with _**no**_ sliced prediction: **`0.607 mAP50`**
3. SAHI with sliced prediction in an oversized slice-size (8000x8000): **`0.607 mAP50`**
4. SAHI with sliced prediction in an appropriate slice-size (4032x4032): **`0.931 mAP50`**

Specifically, I get the same result for 2. and 3., in terms of mAP, Precision, and Recall. And both of them do inference once for each image. 

I pick a picture where the difference is obvious in four scenarios. (I've only picked the ones with significant differences. That doesn't mean all the images are different like this for 2. and 3.)

original image:
![1759_original](https://github.com/obss/sahi/assets/61639889/0c80c74d-6fad-41db-b97a-e2f3d4cb18e0)

label image:
![1759_label](https://github.com/obss/sahi/assets/61639889/3985c19d-a70e-41b7-ba50-a2b4ebcec357)

standard-YOLOv8 image:
![1759_stdyolov8](https://github.com/obss/sahi/assets/61639889/53607d1b-3fd4-439a-821e-7d42ea47835b)

SAHI image with _no_ sliced prediction:
![1759_noslicedSAHI](https://github.com/obss/sahi/assets/61639889/cda347fb-2baf-4321-a0dc-43cfc11ce947)

I'm not sure if the 4 high-resolution images will work on this page......

Here is my code using SAHI. (I do not change any original SAHI codes)

```
source_image_dir = "VOC2007/images/"
dataset_json_path='VOC2007/coco_labels/coco_val.json'
output_dir = 'output'
output_name = 'val' # output_dir/output_name/
yolov8_model_path = "runs/detect/train4/weights/best.pt"
model_type = "yolov8"
model_device = "cuda:0" # or 'cpu'
model_confidence_threshold = 0.001

slice_height = 8000
slice_width = 8000
overlap_height_ratio = 0.2
overlap_width_ratio = 0.2
model_confidence_threshold = 0.001

detection_model = AutoDetectionModel.from_pretrained(
    model_type=model_type,
    model_path=yolov8_model_path,
    confidence_threshold=model_confidence_threshold,
    image_size=640,
    device=model_device,
)
predict(
    model_type=model_type,
    model_path=yolov8_model_path,
    model_device=model_device,
    model_confidence_threshold=model_confidence_threshold,
    source=source_image_dir,
    image_size=640,
    slice_height=slice_height,
    slice_width=slice_width,
    overlap_height_ratio=overlap_height_ratio,
    overlap_width_ratio=overlap_width_ratio,
    postprocess_match_metric='IOU',
    postprocess_match_threshold=0.5,
    dataset_json_path=dataset_json_path,
    visual_bbox_thickness=2,
    project=output_dir,
    name=output_name,
    novisual= True, 
    no_sliced_prediction = True
)
```

Looking forward to your reply.

Best wishes!

**Reply by bshakya77** [reply]
Hello @WaderLaken, 

I was trying to evaluate the prediction results using the sahi cmd. Could you please provide how to performed the evaluation in terms of mAP. When I used the SAHI cmd I got negative mAP, do we need to convert the result.json file to COCO format or is it already generated in COCO format.
I would be thankful if you could include the evaluation code you used.

![evaluation_result](https://github.com/user-attachments/assets/9cb71bd4-746a-4554-9c84-6bf74f173456)

Thank you.

Regards,
Bijay

**Reply by WaderLaken** [reply]
Hi @bshakya77 ,

Here is the SAHI evaluation cmd that I used. 

`python sahi/scripts/coco_evaluation.py /data_bases/data/VOC2007/coco_labels/coco_val.json output/val27/result.json ./`

Specifically, the `coco_evaluation.py` is the official implementation of SAHI for evaluation , and the `coco_val.json` is the coco-format labels of my dataset, and the `result.json`  is the official coco-format predictions of SAHI (generated from `sahi.predict.predict`). I did not  do any changes for these official codes. By the way, the whole code generating `result.json`  had already been shown in front before.

Some suggestions are as follows:

1. Perhaps, you could check the coco-format labels of your dataset. Dose its img_id and img_name corrospond to SAHI prediction ones or is the format of boxes right?
2. Another try is that you could set the appropriate SAHI arguments. Because different slice-size makes significant distinction. 

Best wishes!
Laken

**Reply by bshakya77** [reply]
Thank you for your response. I will try this in my environment.

**Reply by bshakya77** [reply]
Hello @WaderLaken,

I tried your cmd with my dataset.json file and result.json file. But the result came out same. I used the following command:

`!python3 sahi/scripts/coco_evaluation.py datasets/coco128/annotations/instances_val2017.json runs/predict/exp11/result.json ./`

The files I used are attached below. When I checked the format, it looks like both are in COCO dataset format. However, I found difference in the `category_id` in two files for the same `image_id`. May be it is the reason for getting same result.

[result (1).json](https://github.com/user-attachments/files/18382663/result.1.json)
[instances_val2017 (2).json](https://github.com/user-attachments/files/18382664/instances_val2017.2.json)


Is there any way to convert the file in the same category_id format?

Regards,
Bijay

**Reply by fcakyon** [reply]
@bshakya77 see my answer here: https://github.com/obss/sahi/discussions/336#discussioncomment-11817636

**Comment by fcakyon** [comment, marked as answer]
@WaderLaken there was a bug in YOLOv8 prediction thresholds. This PR fixed the issue: https://github.com/obss/sahi/pull/988
